workerbee/postgresql.py

```diff
@@ -28,257 +34,328 @@ def check_name(table_name):
 
 
-def set_job_as_claimed(db_handle, table_id, job_id):
-    db_handle.run(SET_ROW_CLAIMED_BY_ID_QUERY.format(table_id),
-                  parameters={'job_id': job_id})
+def set_job_as_complete(db_handle, tbl_name, job_id, duration,
+                        output_data=None):
+    if output_data is not None:
+        output_data = postgres_jsonify(output_data)
+    db_handle.run(SET_ROW_COMPLETED_BY_ID_QUERY.format(tbl_name=tbl_name),
+                  parameters={'id': job_id, 'job_duration': duration,
+                              'output_data': output_data})
 
 
-def get_total_job_count(db_handle, table_id):
-    return db_handle.one(TOTAL_ROWS_QUERY.format(table_id))
+def set_job_as_claimed(db_handle, tbl_name, job_id):
+    db_handle.run(SET_ROW_CLAIMED_BY_ID_QUERY.format(tbl_name=tbl_name),
+                  parameters={'id': job_id})
 
 
-def get_completed_job_count(db_handle, table_id):
-    return db_handle.one(COMPLETED_ROWS_QUERY.format(table_id))
+def update_job_n_failed_attempts(db_handle, tbl_name, job_id):
+    db_handle.run(UPDATE_ROW_N_FAILED_ATTEMPTS_BY_ID_QUERY.format(tbl_name=tbl_name),
+                  parameters={'id': job_id})
+
+
+def get_total_job_count(db_handle, tbl_name):
+    return db_handle.one(TOTAL_ROWS_QUERY.format(tbl_name=tbl_name))
+
+
+def get_completed_job_count(db_handle, tbl_name):
+    return db_handle.one(COMPLETED_ROWS_QUERY.format(tbl_name=tbl_name))
 
 ################################################################################
 
 
-def postgres_experiment(experiment_id, job_function,
-                        job_ids=None, job_data=None,
-                        host=None, port=None, user=None, database=None,
-                        verbose=True):
+def add_job(experiment_id, input_data, unique_id=None, db_connection_info=None,
+            logger_name=DEFAULT_LOGGER, cursor=None):
+    query_str = INSERT_JOB_QUERY.format(tbl_name=experiment_id)
+    params = {'parameters': {'unique_id': unique_id,
+                             'input_data': postgres_jsonify(input_data)}}
+    if cursor is None:
+        db_handle = get_db_handle(db_info=db_connection_info,
+                                  logger_name=logger_name)
+
+        if not table_exists(db_handle, experiment_id):
+            raise ValueError("Table does not exist for experiment '{}'".format(
+                experiment_id))
+
+        db_handle.run(query_str, **params)
+    else:
+        cursor.run(query_str, **params)
+
+
+def add_jobs(experiment_id, input_datas, unique_ids=None,
+             db_connection_info=None, logger_name=DEFAULT_LOGGER, cursor=None):
+    logger = logging.getLogger(logger_name)
+    if cursor is None:
+        db_handle = get_db_handle(db_info=db_connection_info,
+                                  logger_name=logger_name)
+
+        if not table_exists(db_handle, experiment_id):
+            raise ValueError("Table does not exist for experiment '{}'".format(
+                experiment_id))
```

Comment:
This is undoubtedly a less ambiguous API, but I did at first try:
```py
from workerbee.postgres import add_jobs
add_jobs('my_new_exp', [{'path': 'thepath'}])
```
before realising I needed:
```py
from workerbee.postgres import create_experiment
create_experiment('my_new_exp', [{'path': 'thepath'}])
```
this then feels a little weird, because `create_experiment` is like `add_jobs` only it additionally creates the table (i.e. it still takes data). Re-running `create_experiment` with different job data silently fails to add the new job data, only raising a warning about the table already existing.

I think it might be better to commit further in one explicit/implicit direction here. Either:

1. You have to first call `create_experiment` to instantiate a table for the experiment. Does not take data to add. Calling again is a `ValueError`. `add_jobs`/`add_job` behave as they do now.
 - + Explicit and clear API.
 - - Two function calls needed for common case of _make a new table and add some data_.

2. `create_experiment` goes away - `add_job`/`add_jobs` just implicitly create a table if needed.
  - + Optimises for common case - if the key matches, just ensure that exp exists and add jobs. What's the use of building a table and not putting jobs in it?
  - - Implicit table creation - if you change `experiment_id` by accident you would be highly miffed.